### PR TITLE
[Bugfix] Cannot get large files from PFS

### PR DESCRIPTION
**Pull request by bbonenfant**
This fixes an issue where a users could not retrieve more than 20MiB of a file using `client.get_file`. This also creates a working implementation for `client.get_file_tar`.

### CODE REVIEW

File: tests/experimental/test_pfs.py [hunk 1]
```diff
@@ -773,3 +734,138 @@ def test_mount():
...
+        assert streamed_data[:10] == data[:10]
+        assert streamed_data[-10:] == data[-10:]
+        assert len(streamed_data) == len(data)
+
+    @staticmethod
+    def test_buffered_data(client: Client, repo: str):
+        """Test that data gets buffered as expected."""
+        # Arrange
+        stream_items = [b"a", b"bc", b"def", b"ghij", b"klmno"]
+        stream = (BytesValue(value=item) for item in stream_items)
```
Comment (smalyala) [inline tests/experimental/test_pfs.py:785]: This parenthesis syntax works?

Comment (bbonenfant) [reply tests/experimental/test_pfs.py:785]: Yep, it's the comprehension-syntax for iterators, basically just a lazy list comprehension

Comment (smalyala) [reply tests/experimental/test_pfs.py:785]: Sick. I didn't know generators had their own comprehension too.

File: src/python_pachyderm/mixin/pfs.py [hunk 2]
```diff
@@ -111,11 +71,28 @@ def read(self, size: int = -1) -> bytes:
         bytes
             Content from the stream.
         """
-        return self._file.read(size)
+        try:
+            if size == -1:
+                # Consume the entire stream.
+                for message in self._stream:
```
Comment (smalyala) [inline src/python_pachyderm/mixin/pfs.py:77]: Assuming an infinite stream (or absurdly large stream), should we cap how much we read in before returning so this isn't stuck in an infinite loop?

Comment (bbonenfant) [reply src/python_pachyderm/mixin/pfs.py:77]: That's a good question, I would say that it's fine and I'd rather just be consistent with how `.read()` works with standard python file objects.

Also fun fact, I tried this to see if python would let me do it and I just hard crashed my computer running out of memory:
```python
with open("/dev/urandom") as evil:
    evil.read()
```

File: src/python_pachyderm/mixin/pfs.py [hunk 3]
```diff
@@ -1,60 +1,31 @@
...
-                else:
-                    buf.append(b)
-
-                remaining -= len(buf[-1])
-        except StopIteration:
-            pass
-
-        return b"".join(buf)
+class PfsTarFile(tarfile.TarFile):
+    def __iter__(self):
```
Comment (smalyala) [inline src/python_pachyderm/mixin/pfs.py:20]: Is this called during `extractall()`?

**Review by smalyala — APPROVED** [review]
Left few comments. Really good work with the tests!

File: src/python_pachyderm/mixin/pfs.py (hunk 3 above)
Comment (bbonenfant) [reply src/python_pachyderm/mixin/pfs.py:20]: Yeah, the inner working of `tarfile.TarFile` is kind of strange, it iterates over itself to get the individual member of the TAR. This seemed like the easiest way to insert custom code into that

Comment (smalyala) [reply src/python_pachyderm/mixin/pfs.py:20]: Solid

**Pull request merged**